MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -1,118 +1,182 @@
-                                                             kMSLogTableName, kMSGroupIdColumnName, kMSDataColumnName];
+  NSString *createLogTableQuery = [NSString
+      stringWithFormat:
+          @"CREATE TABLE IF NOT EXISTS %@ (%@ INTEGER PRIMARY KEY AUTOINCREMENT, %@ TEXT NOT NULL, %@ TEXT NOT NULL);",
+          kMSLogTableName, kMSIdColumnName, kMSGroupIdColumnName, kMSDataColumnName];
   [self.connection executeQuery:createLogTableQuery];
 }
 
-#pragma mark - Public
+#pragma mark - Save logs
 
 - (BOOL)saveLog:(id<MSLog>)log withGroupId:(NSString *)groupId {
   if (!log) {
     return NO;
   }
   NSData *logData = [NSKeyedArchiver archivedDataWithRootObject:log];
   NSString *base64Data = [logData base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithLineFeed];
-  NSString *addLogQuery = [NSString
-      stringWithFormat:@"insert or replace into %@ values ('%@', '%@')", kMSLogTableName, groupId, base64Data];
+  NSString *addLogQuery =
+      [NSString stringWithFormat:@"INSERT INTO %@ ('%@', '%@') VALUES ('%@', '%@')", kMSLogTableName,
+                                 kMSGroupIdColumnName, kMSDataColumnName, groupId, base64Data];
   return [self.connection executeQuery:addLogQuery];
 }
 
-- (NSArray<MSLog> *)deleteLogsForGroupId:(NSString *)groupId {
-  NSArray<MSLog> *logs = [self getLogsWithGroupId:groupId];
-  [self deleteLogsWithGroupId:groupId];
-  return logs;
-}
-
-- (void)deleteLogsForId:(__attribute__((unused)) NSString *)batchId withGroupId:(NSString *)groupId {
-
-  // FIXME: Restore batch deletion.
-  [self deleteLogsWithGroupId:groupId];
-}
+#pragma mark - Load logs
 
-- (BOOL)loadLogsForGroupId:(NSString *)groupId
-                     limit:(NSUInteger)limit
-            withCompletion:(nullable MSLoadDataCompletionBlock)completion {
+- (BOOL)loadLogsWithGroupId:(NSString *)groupId
+                      limit:(NSUInteger)limit
+             withCompletion:(nullable MSLoadDataCompletionBlock)completion {
 
   /*
    * There is a need to determine if there will be more logs available than those under the limit.
-   * So this is just about knowing if there is at least 1 log above the limit.
-   * Thus, just 1 log is added to the requested limit and then removed later as needed.
+   * This is just about knowing if there is at least 1 log above the limit.
    */
-  NSMutableArray<MSLog> *logs = (NSMutableArray<MSLog> *)[self getLogsWithGroupId:groupId limit:(limit + 1)];
+  NSMutableDictionary<NSString *, id<MSLog>> *logs =
+      [[self getLogsFromDBWithGroupId:groupId limit:(limit + 1)] mutableCopy];
+  BOOL logsAvailable = NO;
   BOOL moreLogsAvailable = NO;
+  NSString *batchId;
 
-  // Remove the log in excess, it means there is more logs available.
-  if (logs.count > limit) {
-    [logs removeLastObject];
+  // More logs available for the next batch, remove the log in excess for this batch.
+  if (logs.count > 0 && logs.count > limit) {
```

Comment:
It's an optimization, this info is used by the consumer of the logs, if it knows that there is no next batch then no need  to trigger any timer to handle the next batch.